File: verifiers/ts/src/commitment.ts
import { scrypt } from '@noble/hashes/scrypt'
import { verifyStrict } from './ed25519.js'

const LABEL_COMMITMENT = 'Attest-buyer-commitment-v1'
const LABEL_CHALLENGE = 'Attest-binding-challenge-v1'
const SCRYPT = { N: 32768, r: 8, p: 1, dkLen: 32 } as const // fixed, never configurable
const IDENTIFIER_TYPES = new Set(['issuer-account', 'email'])

function stripAsciiWs(s: string): string {
  return s.replace(/^[ \t\n\r]+/, '').replace(/[ \t\n\r]+$/, '')
}

export function normalizeIdentifier(identifier: string, identifierType: string): string {
  if (!IDENTIFIER_TYPES.has(identifierType))
    throw new Error(`unknown identifier_type: '${identifierType}'`)
  let norm: string
  if (identifierType === 'email') {
    norm = stripAsciiWs(identifier).normalize('NFC')
      .replace(/[A-Z]/g, (c) => String.fromCharCode(c.charCodeAt(0) + 32))
  } else {
    norm = identifier.normalize('NFC') // issuer-account: NFC only, case preserved
  }
  if (hasLoneSurrogate(norm))
    // JS TextEncoder would map a lone surrogate to U+FFFD, so distinct identifiers
    // could derive the same commitment; fail closed for parity with Python
    // (2026-07-13 review, finding 6).
    throw new Error('normalized identifier must not contain lone surrogate code points')
  if (norm.includes('\x00')) throw new Error('normalized identifier must not contain 0x00')
  return norm
}

function hasLoneSurrogate(s: string): boolean {
  for (let i = 0; i < s.length; i++) {
    const cp = s.charCodeAt(i)
    if (cp >= 0xd800 && cp <= 0xdbff) {
      const lo = s.charCodeAt(i + 1)
      if (lo >= 0xdc00 && lo <= 0xdfff) { i++; continue }
      return true
    }
    if (cp >= 0xdc00 && cp <= 0xdfff) return true
  }
  return false
}

function concatBytes(...parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((n, p) => n + p.length, 0)
  const out = new Uint8Array(total)
  let o = 0
  for (const p of parts) { out.set(p, o); o += p.length }
  return out
}

export function computeCommitment(identifier: string, identifierType: string, salt: Uint8Array): Uint8Array {
  if (salt.length !== 16) throw new Error('salt must be exactly 16 raw bytes')
  const norm = normalizeIdentifier(identifier, identifierType)
  const enc = new TextEncoder()
  const password = concatBytes(
    enc.encode(LABEL_COMMITMENT), Uint8Array.of(0),
    enc.encode(identifierType), Uint8Array.of(0),
    enc.encode(norm),
  )
  return scrypt(password, salt, SCRYPT)
}

export function verifyChallenge(receiptId: string, nonce: Uint8Array, sig: Uint8Array, pub: Uint8Array): boolean {
  if (nonce.length < 16) throw new Error('nonce must be at least 16 bytes')
  const enc = new TextEncoder()
  const msg = concatBytes(enc.encode(LABEL_CHALLENGE), Uint8Array.of(0), enc.encode(receiptId), Uint8Array.of(0), nonce)
  return verifyStrict(msg, sig, pub)
}

File: src/opr/commitment.py
"""Buyer binding: scrypt commitment and Ed25519 challenge-response (§3.2)."""

from __future__ import annotations

import hashlib
import unicodedata

from opr import keys

LABEL_COMMITMENT = b"OPR-buyer-commitment-v1"
LABEL_CHALLENGE = b"OPR-binding-challenge-v1"
IDENTIFIER_TYPES = ("issuer-account", "email")

_SCRYPT_N = 32768
_SCRYPT_R = 8
_SCRYPT_P = 1
_SCRYPT_DKLEN = 32
_SCRYPT_MAXMEM = 64 * 1024 * 1024
_ASCII_WS = " \t\n\r"


def normalize(identifier: str, identifier_type: str) -> str:
    if identifier_type not in IDENTIFIER_TYPES:
        raise ValueError(f"unknown identifier_type: {identifier_type!r}")
    if identifier_type == "email":
        identifier = identifier.strip(_ASCII_WS)
        identifier = unicodedata.normalize("NFC", identifier)
        identifier = "".join(chr(ord(c) + 32) if "A" <= c <= "Z" else c for c in identifier)
    else:  # issuer-account: NFC only, exact string otherwise
        identifier = unicodedata.normalize("NFC", identifier)
    if "\x00" in identifier:
        raise ValueError("normalized identifier must not contain 0x00")
    return identifier


def compute(identifier: str, identifier_type: str, salt: bytes) -> bytes:
    if len(salt) != 16:
        raise ValueError("salt must be exactly 16 raw bytes")
    norm = normalize(identifier, identifier_type)
    password = LABEL_COMMITMENT + b"\x00" + identifier_type.encode() + b"\x00" + norm.encode()
    return hashlib.scrypt(
        password,
        salt=salt,
        n=_SCRYPT_N,
        r=_SCRYPT_R,
        p=_SCRYPT_P,
        maxmem=_SCRYPT_MAXMEM,
        dklen=_SCRYPT_DKLEN,
    )


def challenge_message(receipt_id: str, nonce: bytes) -> bytes:
    if len(nonce) < 16:
        raise ValueError("nonce must be at least 16 bytes")
    return LABEL_CHALLENGE + b"\x00" + receipt_id.encode() + b"\x00" + nonce


def sign_challenge(receipt_id: str, nonce: bytes, kp: keys.SigningKeyPair) -> bytes:
    return keys.sign(challenge_message(receipt_id, nonce), kp)


def verify_challenge(receipt_id: str, nonce: bytes, sig: bytes, pub: bytes) -> bool:
    return keys.verify_strict(challenge_message(receipt_id, nonce), sig, pub)
